elasticsearch/store/embedding_service.py

```diff
@@ -0,0 +1,132 @@
+import asyncio
+from abc import ABC, abstractmethod
+from typing import List, Optional
+
+import nest_asyncio  # type: ignore
+from elasticsearch import AsyncElasticsearch
+
+from elasticsearch.store._utilities import create_elasticsearch_client
+
+
+class EmbeddingService(ABC):
+    @abstractmethod
+    async def embed_documents_async(self, texts: List[str]) -> List[List[float]]:
+        """Generate embeddings for a list of documents.
+
+        Args:
+            texts: A list of document strings to generate embeddings for.
+
+        Returns:
+            A list of embeddings, one for each document in the input.
+        """
+
+    @abstractmethod
+    def embed_documents(self, texts: List[str]) -> List[List[float]]:
+        """Generate embeddings for a list of documents.
+
+        Args:
+            texts: A list of document strings to generate embeddings for.
+
+        Returns:
+            A list of embeddings, one for each document in the input.
+        """
+
+    @abstractmethod
+    async def embed_query_async(self, query: str) -> List[float]:
+        """Generate an embedding for a single query text.
+
+        Args:
+            text: The query text to generate an embedding for.
+
+        Returns:
+            The embedding for the input query text.
+        """
+
+    @abstractmethod
+    def embed_query(self, query: str) -> List[float]:
+        """Generate an embedding for a single query text.
+
+        Args:
+            text: The query text to generate an embedding for.
+
+        Returns:
+            The embedding for the input query text.
+        """
+
+
+class ElasticsearchEmbeddings(EmbeddingService):
```

Comment:
I like this, really declutters the code I work with.

One question: are test cases also handled, including `@pytest.mark.asyncio` markers?

Reply:
Great question! It's not handled, because `unasync` only handles single tokens. See https://github.com/elastic/elasticsearch-dsl-py/pull/1714#discussion_r1527918832 for the same discussion in the DSL client. For now we can enable pytest-asyncio's auto mode.

Reply:
I discussed quickly with @miguelgrinberg and actually it's best to handle this with sed, as we do to remove `import asyncio` and `asyncio.run()` from DSL examples: https://github.com/elastic/elasticsearch-dsl-py/blob/f56d3a52278e3caff045bfce320415526c56bb2d/utils/run-unasync.py#L98-L109.

Reply:
Two more questions:
1. What about Generator type annotations? `AsyncGenerator[A,B]` takes 2 types and `Generator[A, B, C]` takes 3. It seems this is not handled. 
2. It seems that import paths are not adjusted or maybe I'm doing something wrong. `from elasticsearch.vectorstore._async._utils import foo` remains the same.

Reply:
Findings:
1. Can be solved by using `AsyncIterator[A]`.
2. Can be solved through `additional_replacements`.
3. I guess the import hurts noone.

Reply:
1. Good if AsyncIterator works for you.
2. Correct, see https://github.com/elastic/elasticsearch-dsl-py/blob/b5800d48139c99ca5de3540f74f7863c8d51adf9/utils/run-unasync.py#L51
3. It will probably fail lint, you could use sed to remove it alongside the pytest-asyncio decorators.

Reply:
@maxjakob I have used unasync to do the async conversion of the DSL client recently. As you noticed there are some things that unasync does not handle, and for the DSL client I have added an extra pass that addresses these using `sed`. Happy to discuss the details if it is helpful.

Reply:
@miguelgrinberg I'm already heavily copying from you :) see the changes in run-unasync